Add golangci lint integration.

jpeach opened the pull request:
Add golangci-lint integration and fix or ignore all the resulting issues. In general, the things I ignored were names that can't be changed without a deprecation and migration.

This updates #414.

jpeach commented:
Rebased.

alecholmez commented on internal/example/main/main.go:
```diff
@@ -25,12 +25,8 @@ import (
-
-	port     uint
-	basePort uint
-	mode     string
```
was this unused?

jpeach replied:
Yup.

alecholmez commented on pkg/cache/v3/cache.go:
```diff
@@ -186,7 +186,7 @@ func (r *RawResponse) GetDiscoveryResponse() (*discovery.DiscoveryResponse, erro
 
 		for i, resource := range r.Resources {
-			maybeTtldResource, resourceType, err := r.maybeCreateTtlResource(resource)
+			maybeTtldResource, resourceType, err := r.maybeCreateTTLResource(resource)
```
I'd rather keep the TTL stuff consistent. Maybe change `ResourceWithTtl` to `ResourceWithTTL`?

jpeach replied:
This would be `TTL` according to Go's [initialism convention](https://github.com/golang/go/wiki/CodeReviewComments#initialisms). I updated the internal names, but I didn't want to change the API in this PR.

alecholmez replied:
Yup I agree. I honestly am okay with changing the API. If that breaks users they can pin, and we also haven't hit `v1.0` yet.

jpeach replied:
Ok, changing it. AFAIK we can't add a compatibility alias to a field so there's no way to rename without a breaking change.

alecholmez commented on pkg/cache/v3/delta_test.go:
```diff
@@ -175,24 +175,21 @@ func TestConcurrentSetDeltaWatch(t *testing.T) {
 						ResourceNamesSubscribe: []string{clusterName},
 					}, stream.NewStreamState(false, make(map[string]string)), responses)
+
+					defer cancel()
```
Did this not cause test failures?

jpeach replied:
Nope, it was fine :)

alecholmez replied:
Nice, that was causing issues for me a while ago but I guess it's fixed

alecholmez commented on pkg/cache/v3/mux.go:
```diff
@@ -37,6 +37,16 @@ type MuxCache struct {
+	// a Mutex in order to trigger this check (i.e. not for actually
+	// locking fields).
+	//
+	// So in this specific case, it happens to be safe to copy the
```
I would not be opposed to changing the API to not copy lock values. We did this for delta.

jpeach replied:
OK, then I'll change this to take a pointer.

alecholmez commented on pkg/cache/v3/simple.go:
```diff
@@ -154,7 +150,7 @@ func NewSnapshotCacheWithHeartbeating(ctx context.Context, ads bool, hash NodeHa
 }
 
-func (cache *snapshotCache) sendHeartbeats(ctx context.Context, node string) {
+func (cache *snapshotCache) sendHeartbeats(_ context.Context, node string) {
```
Were there linters complaining about this? I generally think this is bad to have in param naming schemes

jpeach replied:
IIRC this is an unused variable warning. Perhaps I can just remove the parameter altogether.

alecholmez replied:
I'd remove the param

alecholmez commented on pkg/cache/v3/snapshot.go:
```diff
@@ -66,15 +66,15 @@ func IndexRawResourcesByName(items []types.Resource) map[string]types.Resource {
 
 // NewResources creates a new resource group.
-func NewResourcesWithTtl(version string, items []types.ResourceWithTtl) Resources {
+func NewResourcesWithTtl(version string, items []types.ResourceWithTtl) Resources { // nolint:golint,revive
```
Consistency with the `TTL` casing

jpeach replied:
So Go naming conventions really want this to be `TTL`, but I didn't want to break any compatibility in this PR. Since this is a public function, I just annotated it.

Maybe rename and deprecate?

```Go
// Deprecated 
func NewResourcesWithTtl(version string, items []types.ResourceWithTtl) Resources {
        return NewResourcesWithTTL(...)
}

func NewResourcesWithTTL(version string, items []types.ResourceWithTtl) Resources {
...
}
```

alecholmez replied:
I'm all for the rename. Lets make it consistent throughout

jpeach replied:
AFAIK we can't add compatibility aliases since we need to rename the `Ttl` field in `ResourceWithTtl` and there's no way to alias that.

alecholmez commented on pkg/cache/v3/snapshot.go:
```diff
@@ -137,9 +137,9 @@ func NewSnapshotWithResources(version string, resources SnapshotResources) Snaps
 }
 
-type ResourceWithTtl struct {
+type ResourceWithTtl struct { // nolint:golint,revive
```
Same here

jpeach replied:
Removed this, since it is unused in this repository.

alecholmez commented on pkg/resource/v3/resource.go:
```diff
@@ -43,7 +43,7 @@ func GetHTTPConnectionManager(filter *listener.Filter) *hcm.HttpConnectionManage
 	// use typed config if available
 	if typedConfig := filter.GetTypedConfig(); typedConfig != nil {
-		ptypes.UnmarshalAny(typedConfig, config)
+		_ = ptypes.UnmarshalAny(typedConfig, config)
```
Why the value ignore?

jpeach replied:
IIRC this is errcheck again :)

alecholmez replied:
Hm, we probably shouldn't ignore this error. Maybe we should alter the function signature and bubble up that err

jpeach replied:
Updated to return nil which I this is OK for this API.

alecholmez commented on pkg/server/v3/gateway_test.go:
```diff
@@ -100,7 +100,7 @@ func TestGateway(t *testing.T) {
 			t.Fatal(err)
 		}
-		resp, code, err := gtw.ServeHTTP(req)
+		resp, code, _ := gtw.ServeHTTP(req)
```
Let's not ignore the errors here, I'd rather add an error check

alecholmez commented on pkg/server/v3/gateway_test.go:
```diff
@@ -114,7 +114,7 @@ func TestGateway(t *testing.T) {
 			t.Fatal(err)
 		}
-		resp, code, err := gtw.ServeHTTP(req)
+		resp, code, _ := gtw.ServeHTTP(req)
```
Same

alecholmez commented on pkg/test/main/main.go:
```diff
@@ -307,7 +311,7 @@ func callEcho() (int, int) {
 				return
 			}
-			defer req.Body.Close()
+			defer func() { _ = req.Body.Close() }()
```
More unnamed values

alecholmez commented on pkg/test/server.go:
```diff
@@ -111,7 +111,7 @@ func RunManagementGateway(ctx context.Context, srv server.Server, port uint, lg
 
 	// Cleanup our gateway if we receive a shutdown
-	server.Shutdown(ctx)
+	_ = server.Shutdown(ctx)
```
Unnamed value

jpeach replied:
Added a log to match the error on listen.

alecholmez reviewed (changes requested):
Thanks for the changes and fixing the lint issues! Just a few comments for you, I'd rather stay away from the unnamed values and params especially in situations where theres no need for a partial value read or something of that nature.

alecholmez commented:
I think @snowp or @mattklein123 will need to change CI to run the linter. I don't have those permissions

alecholmez commented:
Just FYI I'm more in favor of adding extra error checks and changing things around if needed to support that. I'd rather not ignore things

jpeach commented:
@alecholmez  this dropped off my list a bit. Any objections to me rebasing while I address your review feedback?

jpeach commented:
Addresses all the review feedback. Note that this includes incompatible changes, so users of some APIs will need to make minor changes when they upgrade.

alecholmez commented:
@jpeach sorry for this, can you rebase? I'd like to get this in before we attempt to cut a `v1` release. Want to make sure the code quality is up to par

jpeach commented:
> @jpeach sorry for this, can you rebase? I'd like to get this in before we attempt to cut a `v1` release. Want to make sure the code quality is up to par

Rebased

If you are planning a v1 release, I'd recommend you make a pass and s/Ttl/TTL/ in all the public names. I fixed some that linters picked up, but better to make it all consistent now (and in a separate PR).

alecholmez reviewed (approved):
Thanks for this! I will try and knock off some more lint issues as we move towards `v1`
